### PR TITLE
Updating live tests

### PR DESCRIPTION
**Pull request by justinThompson**
- updated tests but the 'testLoadBalancerWithTcpTargets' test is still broken because of a bug in the version of netcat installed on the server

### CODE REVIEW

**Comment by justinThompson** [comment]
i tested with a later version of the centos AMI (RightImage_CentOS_6.3_x64_v5.8.8.8_EBS - ami-eb0b8082) but the netcat problem remains

File: src/main/java/brooklyn/entity/proxy/aws/ElbController.java [hunk 1]
```diff
@@ -93,10 +93,7 @@
     
     // http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/elasticloadbalancing/model/HealthCheck.html#setTarget(java.lang.String)
     ConfigKey<String> HEALTH_CHECK_TARGET = ConfigKeys.newStringConfigKey(
-            "aws.elb.healthCheck.target", "Specifies the instance being checked", "${instanceProtocol}:${instancePort?c}/");
-    
-    ConfigKey<Boolean> HEALTH_CHECK_ENABLED = ConfigKeys.newBooleanConfigKey(
-            "aws.elb.healthCheck.enabled", "Whether to do health checks for the instances", true);
+            "aws.elb.healthCheck.target", "The instance being checked. The protocol is either TCP, HTTP, HTTPS, or SSL. The range of valid ports is one (1) through 65535.", "HTTP:8080/");
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbController.java]: ` The range of valid ports is one (1) through 65535` - can omit this as a general port restriction; if you decide to keep it - looks a bit weird, could change to "The range of valid ports is 1-65535".

File: pom.xml [hunk 2]
```diff
@@ -6,6 +6,10 @@
     <version>0.4.0-SNAPSHOT</version> <!-- BROOKLYN_AWS_ELB_VERSION -->
     <packaging>jar</packaging>
 
+    <properties>
+        <jclouds.labs.version>2.0.1-SNAPSHOT</jclouds.labs.version>
```
Comment (neykov) [inline pom.xml]: Current snapshot is `2.1.0-SNAPSHOT`

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 3]
```diff
@@ -58,6 +58,9 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(ElbControllerImpl.class);
 
+    private ELBApi elbApi;
+    private LoadBalancerServiceContext loadBalancerServiceContext;
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Looking at the code it made sense to have these as local variables. Why did you move them as class fields?

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 4]
```diff
@@ -111,22 +114,20 @@ public void reload() {
...
             Set<String> targetAddresses = super.getServerPoolAddresses();
 
-            LoadBalancerServiceContext loadBalancerServiceContext = null;
             try {
                 loadBalancerServiceContext = newLoadBalancerServiceContext(loc.getIdentity(), loc.getCredential());
-                ELBApi api = api(loadBalancerServiceContext);
 
                 Set<String> instances = ImmutableSet.copyOf(targetAddresses);
-                LoadBalancer loadBalancer = api.getLoadBalancerApi().get(elbName);
+                LoadBalancer loadBalancer = getELBApi().getLoadBalancerApi().get(elbName);
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: `getELBApi` is initialised only once, whereas `loadBalancerServiceContext` is created in each method and then destroyed (closed). Seems wrong to re-use the same `getELBApi` if the context it's been created off is already destroyed.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 5]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
-                            Listener.builder()
-                                    .protocol(Protocol.HTTP)
-                                    .instancePort(instancePort)
-                                    .port(loadBalancerPort)
-                                    .build(),
-                            availabilityZoneNames);
+            String elbDnsName = "";
+            Listener listener = buildListener();
+//          create vanilla load balancer
+            if(availabilityZoneNames != null && availabilityZoneNames.size() > 0 ){
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: `availabilityZoneNames` is always non-null. Same for `subnets` below.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 6]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+                elbDnsName = getELBApi().getLoadBalancerApi().createLoadBalancerInAvailabilityZones(elbName, ImmutableSet.<Listener>of(listener), availabilityZoneNames);
+            }
+            else if(subnets != null && !subnets.isEmpty()) {
+                elbDnsName = getELBApi().getLoadBalancerApi().createLoadBalancerInSubnets(elbName, ImmutableSet.<Listener>of(listener), subnets);
+            }
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java:199]: What if both are empty? Throw an exception. You can remove the initializing empty string on `elbDnsName` then to show that a value is always set (or exception throw).

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 7]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+            configureListeners(elbName, loadbalancer);
 
             sensors().set(Attributes.HOSTNAME, elbDnsName);
         } finally {
             Closeables2.closeQuietly(loadBalancerServiceContext.unwrap());
         }
     }
 
+    private void configureHealthCheck() {
+        String elbName = getConfig(LOAD_BALANCER_NAME);
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Either pass the name everywhere, or fetch it from config everywhere.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 8]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+        final LoadBalancer loadBalancer = getELBApi().getLoadBalancerApi().get(elbName);
+        final HealthCheck healthCheck = loadBalancer.getHealthCheck();
+
+        final HealthCheck healthCheckConfig = HealthCheck.builder()
+                .healthyThreshold(healthCheckHealthyThreshold != null ? healthCheckHealthyThreshold : healthCheck.getHealthyThreshold())
+                .unhealthyThreshold(healthCheckUnhealthyThreshold != null ? healthCheckUnhealthyThreshold : healthCheck.getUnhealthyThreshold())
+                .interval(healthCheckInterval != null ? healthCheckInterval : healthCheck.getInterval())
+                .target(healthCheckTarget != null ? healthCheckTarget : healthCheck.getTarget())
+                .timeout(healthCheckTimeout != null ? healthCheckTimeout : healthCheck.getTimeout())
+                .build();
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java:228]: The builder has `fromListener` method which will initialise it from the `healthCheck` you already have. Then you can just set the non-null values. But this is fine as well.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 9]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+
+            configureAvailabilityZones(elbName, loadBalancer);
+
+            configureSubnets(elbName, loadBalancer);
+
+            configureListeners(elbName, loadBalancer);
+            configureHealthCheck();
+
+            sensors().set(Attributes.HOSTNAME, loadBalancer.getDnsName());
+            }
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Wrong indentation.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 10]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+        finally {
+            Closeables2.closeQuietly(loadBalancerServiceContext.unwrap());
+        }
+
+    }
+    private void configureSecurityGroups(String elbName){
+        final Collection<String> securityGroups = getConfig(LOAD_BALANCER_SECURITY_GROUPS);
+
+        if (securityGroups != null && securityGroups.size() > 0) {
+            getELBApi().getLoadBalancerApi().applySecurityGroupsToLoadBalancer(elbName, securityGroups != null ? securityGroups : ImmutableList.<String>of());
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: `securityGroups` is non-null here.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 11]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+
+        if (availabilityZoneNames != null && availabilityZoneNames.size() > 0) {
+
             // Set the availability zones
             Set<String> oldAvailabilityZoneNames = loadBalancer.getAvailabilityZones();
             Set<String> oldAvailabilityZoneNamesSet = ImmutableSet.copyOf(oldAvailabilityZoneNames);
-            Set<String> removedAvailabilityZoneNames = Sets.difference(oldAvailabilityZoneNamesSet, availabilityZoneNames);
-            Set<String> addedAvailabilityZoneNames = Sets.difference(availabilityZoneNames, oldAvailabilityZoneNamesSet);
+            Set<String> removedAvailabilityZoneNames = Sets.difference(oldAvailabilityZoneNamesSet, ImmutableSet.copyOf(availabilityZoneNames));
+            Set<String> addedAvailabilityZoneNames = Sets.difference(ImmutableSet.copyOf(availabilityZoneNames), oldAvailabilityZoneNamesSet);
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: No point in copying `availabilityZoneNames` again - it's already copied and immutable (from above).

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 12]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
             if (!addedAvailabilityZoneNames.isEmpty()) {
-                api.getAvailabilityZoneApi().addAvailabilityZonesToLoadBalancer(addedAvailabilityZoneNames, elbName);
+                getELBApi().getAvailabilityZoneApi().addAvailabilityZonesToLoadBalancer(addedAvailabilityZoneNames, elbName);
             }
             if (!removedAvailabilityZoneNames.isEmpty()) {
-                api.getAvailabilityZoneApi().removeAvailabilityZonesFromLoadBalancer(addedAvailabilityZoneNames, elbName);
+                getELBApi().getAvailabilityZoneApi().removeAvailabilityZonesFromLoadBalancer(addedAvailabilityZoneNames, elbName);
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: should be `removedAvailabilityZoneNames`

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 13]
```diff
@@ -407,7 +410,7 @@ protected void deleteLoadBalancer(String elbName) {
             }
         }
         return ImmutableSet.copyOf(availabilityZones);
-    }
+    } */
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Delete unused code - can be recovered from source control.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 14]
```diff
@@ -473,11 +480,14 @@ protected boolean isAvailabilityZone(String name) {
 
     protected JcloudsLocation getLocation() {
         JcloudsLocation result = getAttribute(JCLOUDS_LOCATION);
-        checkNotNull(result, "JcloudsLocation not set - was ELB started, or has it been stopped?");
+        //checkNotNull(result, "JcloudsLocation not set - was ELB started, or has it been stopped?");
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Delete

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 15]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+                elbDnsName = getELBApi().getLoadBalancerApi().createLoadBalancerInAvailabilityZones(elbName, ImmutableSet.<Listener>of(listener), availabilityZoneNames);
+            }
+            else if(subnets != null && !subnets.isEmpty()) {
+                elbDnsName = getELBApi().getLoadBalancerApi().createLoadBalancerInSubnets(elbName, ImmutableSet.<Listener>of(listener), subnets);
+            }
+
+            LoadBalancer loadbalancer = getELBApi().getLoadBalancerApi().get(elbName);
+            configureListeners(elbName, loadbalancer);
+            configureHealthCheck();
+            configureListeners(elbName, loadbalancer);
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Double call to `configureListeners`

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 16]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+
+    private void configureListeners(String elbName, LoadBalancer loadBalancer) {
+        Set<ListenerWithPolicies> listenerWithPolicies = loadBalancer.getListeners();
+        Set<Integer> listenerPorts = Sets.newHashSet();
+        for (ListenerWithPolicies listenerWithPolicy : listenerWithPolicies) {
+            listenerPorts.add(listenerWithPolicy.getPort());
         }
 
+        if (!listenerPorts.isEmpty()) {
+            getELBApi().getLoadBalancerApi().deleteLoadBalancerListeners(elbName, listenerPorts);
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Don't see `deleteLoadBalancerListeners` on `LoadBalancerApi` - probably to be PR'ed separately?

File: src/main/java/brooklyn/entity/proxy/aws/ElbController.java (hunk 1 above)
Comment (neykov) [reply src/main/java/brooklyn/entity/proxy/aws/ElbController.java]: Isn't the old default expression working? Better leave it without a default here and build it in the code so it matches the `instanceProtocol`, `instancePort` settings.
Why leave out the enable flag?

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java [hunk 17]
```diff
@@ -173,175 +174,179 @@ protected void startLoadBalancer() {
...
+
+        final HealthCheck healthCheckConfig = HealthCheck.builder()
+                .healthyThreshold(healthCheckHealthyThreshold != null ? healthCheckHealthyThreshold : healthCheck.getHealthyThreshold())
+                .unhealthyThreshold(healthCheckUnhealthyThreshold != null ? healthCheckUnhealthyThreshold : healthCheck.getUnhealthyThreshold())
+                .interval(healthCheckInterval != null ? healthCheckInterval : healthCheck.getInterval())
+                .target(healthCheckTarget != null ? healthCheckTarget : healthCheck.getTarget())
+                .timeout(healthCheckTimeout != null ? healthCheckTimeout : healthCheck.getTimeout())
+                .build();
+
+        getELBApi().getHealthCheckApi().configureHealthCheck(elbName, healthCheckConfig);
```
Comment (neykov) [inline src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Does this overwrite existing health checks?

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 15 above)
Comment (neykov) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Missing call to `configureSecurityGroups`.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 10 above)
Comment (neykov) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: Do you need to fetch the existing security groups if some were applied manually?

**Review by neykov — CHANGES_REQUESTED** [review]

File: src/main/java/brooklyn/entity/proxy/aws/ElbController.java (hunk 1 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbController.java]: yep, ill remove the ports part.

I had problems with the old expression, but ill take another look at it.

the reason i removed the enable flag is because a health check is always created. So even if you say you do not want a healthcheck, a default one will be created.

File: pom.xml (hunk 2 above)
Comment (justinThompson) [reply pom.xml]: will update

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 3 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: the idea was that we could reuse the elbApi if it existed already, rather than having to create it each time, but you think put it back to local variables?

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 4 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: good point, if i change the elbapi back to a local variable (as in the comment above) it will address this

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 5 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: will update

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 6 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java:199]: what is the correct exception to throw for config etc? and set it in the service problems?

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 3 above)
Comment (neykov) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: What concerns me:
  * `loadBalancerServiceContext` is created
  * `elbApi` is created out of `loadBalancerServiceContext`
  * `loadBalancerServiceContext` is destroyed
  * what's the state of `elbApi` now? Don't think we should be using it at this point?

Moreover `loadBalancerServiceContext` is created and destroyed in each method - either create it once and re-use it together with `elbApi` or move them both as local variables.

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 17 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: yes, as far as i'm aware you can only have one health check so it will overwrite it

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 9 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: 👍

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 10 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: fetch the existing security groups to remove them?

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 11 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: 👍

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 12 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: 👍

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 16 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: this is in a separate jcloud labs PR, should i leave this out until that PR is merged?

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 13 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: 👍

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 14 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: 👍

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 3 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java]: ok, that makes sense. it seems safer to use them both as local variables, so ill make that change

File: src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java (hunk 8 above)
Comment (justinThompson) [reply src/main/java/brooklyn/entity/proxy/aws/ElbControllerImpl.java:228]: the reason i did it this way was so if you used the `BIND_TO_EXISTING` option you could reconfigure the health check rather than it just using the configured health check.

**Review by neykov — APPROVED** [review]
